src/handlers/mod.rs

```diff
@@ -419,14 +420,36 @@ impl ForeignToplevelHandler for State {
 delegate_foreign_toplevel!(State);
 
 impl ScreencopyHandler for State {
-    fn frame(&mut self, screencopy: Screencopy) {
-        if let Err(err) = self
-            .niri
-            .render_for_screencopy(&mut self.backend, screencopy)
-        {
-            warn!("error rendering for screencopy: {err:?}");
+    fn frame(&mut self, manager: &Weak<ZwlrScreencopyManagerV1>, screencopy: Screencopy) {
+        let Ok(manager) = manager.upgrade() else {
+            trace!("screencopy manager already destroyed {manager:?}");
+            return;
+        };
+
+        // If with_damage then push it onto the queue for redraw of the output,
+        // otherwise render it immediately.
+        if screencopy.with_damage() {
+            let queue = self.niri.screencopy_state.get_mut(&manager).unwrap();
+            queue.push(screencopy);
+        } else {
+            self.backend.with_primary_renderer(|renderer| {
+                if let Err(err) = self
+                    .niri
+                    .render_for_screencopy(renderer, &manager, screencopy)
+                {
+                    warn!("error rendering for screencopy: {err:?}");
+                }
+            });
         }
     }
+
+    fn bind(&mut self, manager: &ZwlrScreencopyManagerV1) {
+        self.niri.screencopy_state.bind_manager(manager);
+    }
+
+    fn destroy(&mut self, manager: &ZwlrScreencopyManagerV1) {
+        self.niri.screencopy_state.destroy_manager(manager);
+    }
```

Comment:
Instead of making these methods public, usually you add one method `fn screencopy_state(&mut self) -> &mut ScreencopyManager`. Check other protocol impls for an example.

Reply:
I was hoping to restrict the interface so it won't be able to call other `ScreencopyManagerState` public methods, but I guess this is fine too.